content/blog/2019-12-03-posits/index.md

```diff
@@ -0,0 +1,262 @@
+
+Accuracy of single argument operations between representations are compared by plotting sorted error.
+For instance, Gustafson compares accuracy of square root in the following plot:
+
+![](sqrt-losses.png)
+
+
+# Evaluation
+
+As part of our cultist duties, we compare the accuracy of 32-bit floating point and posit representation by comparing their accuracy under a variety of benchmarks.
+In each benchmark, we express real number calculations in terms of operations over 64-bit `double`s.
+In other words, the `double` type serves as the "oracle" baseline from which we compute accuracy.
+Each `double` benchmark is then compiled to LLVM IR, upon which we apply a `posit` LLVM pass and a `float` LLVM pass respectively to generate associated benchmarks.
+
+Finally, we compute percent error by comparing `float` and `posit` benchmark results to the `double` baseline benchmark.
+We regard benchmark errors as arising from the inaccuracy that is characteristic of the particular 32-bit representation, compounded with successive operations.
+With our deepest apologies to Gustafson, we compute these errors on a linear scale, rather than a logarithmic one, and use these as metrics to compare the accuracy of the two representations.
+
+
+## LLVM pass implementation
+
+We aimed for our `float` and `posit` passes to insert as much representation-specific functionality as possible into our benchmarks.
+The case of `float`s allows a full translation, since every `double` operation can be cast as a `float` operations.
+The case of `posit`s lacked full support, however: we did not provide implementations for all existing LLVM `double` operations.
+
+To accommodate both cases under one pass implementation, 
+we designed our pass such that it first lowers each supported `double` operation and operands down to the target representation, subsequently lifting the result back to a `double`.
+
+Since each rewrite has a purely local effect, we can avoid reasoning about complex allocation and type casting scenarios.
+Though computationally inefficient, this approach immediately supports all programs and is easily extendible.
+
+In pseduo-LLVM, the general structure of a pass to type `typ` converts an arbitrary `double` operation `op` and its operands proceeds as follows:
+
+```llvm 
+x : double = ..
+y : double = ..
+result : double = op double x y
+```
+
+-->
+
+
+```llvm
+x : double = ..
+y : double = ..
+
+x_typ : typ = convert double x to typ;
+y_typ : typ = convert double y to typ;
+result_typ : typ = typ_op x_typ y_typ; 
+
+result double = convert result_typ typ to double
+```
+
+This generic pass structure parametrizes `typ` into three components:
+- `convert double to typ`: conversion from `double` to `typ`
+- `convert typ to double`: conversion from `typ` to `double`
+- `typ_op`: implementation over `typ` values of `op`, the corresponding `double` operation
+
+The `float` pass specifies these components as:
+- `fptrunc double to float`
+- `fpext float to double`
```

Comment:
Maybe clarify that these are built-in LLVM instructions?

Reply:
clarified.